### PR TITLE
[mir-opt] Fix `Inline` pass to handle inlining into `box` expressions

### PR DESCRIPTION
**Pull request by wesleywiser**
r? @oli-obk 

Before, the test case just ICE'd here:

https://github.com/rust-lang/rust/blob/a605441e049f0b6d5f7715b94b8ac4662fd7fcf6/src/librustc_mir/transform/inline.rs#L668

### CODE REVIEW

**Comment by oli-obk** [comment]
Would `x.field = foo();` have hit the same ICE before?

File: src/test/mir-opt/inline/inline-into-box-place.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// scope 1 {
+//   debug _x => _1;
+// }
+// scope 2 {
+// }
+// bb0: {
+//   StorageLive(_1);
+//   StorageLive(_2);
+//   _2 = Box(std::vec::Vec<u32>);
+//   _4 = &mut (*_2);
```
Comment (oli-obk) [inline src/test/mir-opt/inline/inline-into-box-place.rs]: Any idea where this reborrow comes from?

**Comment by wesleywiser** [comment]
> Would `x.field = foo();` have hit the same ICE before?

I tried this

```rust
#[inline]
fn bar() -> i32 {
    3
}

fn main() {
    let mut foo = (1, 2);
    foo.1 = bar();
}
```

and it doesn't trigger the issue because we assign the return of `bar()` to a temporary and then assign that temporary to `foo.1` after the call. This ICE occurs specifically when the destination contains a projection.

File: src/test/mir-opt/inline/inline-into-box-place.rs (hunk 1 above)
Comment (wesleywiser) [reply src/test/mir-opt/inline/inline-into-box-place.rs]: I believe it comes from here

https://github.com/rust-lang/rust/blob/a605441e049f0b6d5f7715b94b8ac4662fd7fcf6/src/librustc_mir/transform/inline.rs#L449-L470

**Comment by oli-obk** [comment]
> This ICE occurs specifically when the destination contains a projection.

So... `let x = (42, foo());`? Or how is `box` special in this case?

File: src/test/mir-opt/inline/inline-into-box-place.rs (hunk 1 above)
Comment (oli-obk) [reply src/test/mir-opt/inline/inline-into-box-place.rs]: o.m.g. we really need to get rid of index and deref projections.

**Comment by wesleywiser** [comment]
`box` is special because it causes `dest_needs_borrow()` to return true. This triggers the reborrow and means that when the `Integrator` runs, `destination` has a projection. For `let x = (42, foo());`, `destination` is a plain `Local`.

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/67448#issuecomment-567890066, @bors r-

**Comment by wesleywiser** [comment]
Rebased and modified new test to skip wasm which aborts instead of unwinding.

@bors r=oli-obk

**Pull request merged**